source/particle/property/crystal_preferred_orientation.cc

```diff
@@ -0,0 +1,1102 @@
+                                                            const double /*volume_fraction_mineral*/,
+                                                            const std::array<double,4> & /*ref_resolved_shear_stress*/) const
+      {
+        std::pair<std::vector<double>, std::vector<Tensor<2,3>>> derivatives;
+        switch (cpo_derivative_algorithm)
+          {
+            case CPODerivativeAlgorithm::spin_tensor:
+            {
+              return compute_derivatives_spin_tensor(velocity_gradient_tensor);
+              break;
+            }
+            default:
+              AssertThrow(false, ExcMessage("Internal error."));
+              break;
+          }
+        AssertThrow(false, ExcMessage("Internal error."));
+        return derivatives;
+      }
+
+
+
+      template <int dim>
+      std::pair<std::vector<double>, std::vector<Tensor<2,3>>>
+      CrystalPreferredOrientation<dim>::compute_derivatives_spin_tensor(const Tensor<2,3> &velocity_gradient_tensor) const
+      {
+        // dA/dt = W * A, where W is the spin tensor and A is the rotation matrix
+        // The spin tensor is defined as W = 0.5 * ( L - L^T ), where L is the velocity gradient tensor.
+        const Tensor<2,3> spin_tensor = -0.5 *(velocity_gradient_tensor - dealii::transpose(velocity_gradient_tensor));
+
+        return std::pair<std::vector<double>, std::vector<Tensor<2,3>>>(std::vector<double>(n_grains,0.0), std::vector<Tensor<2,3>>(n_grains, spin_tensor));
+      }
+
+
+
+      template<int dim>
+      unsigned int
+      CrystalPreferredOrientation<dim>::get_number_of_grains()
+      {
+        return n_grains;
+      }
+
+
+
+      template<int dim>
+      unsigned int
+      CrystalPreferredOrientation<dim>::get_number_of_minerals()
+      {
+        return n_minerals;
+      }
+
+
+
+      template <int dim>
+      void
+      CrystalPreferredOrientation<dim>::declare_parameters (ParameterHandler &prm)
+      {
+        prm.enter_subsection("Postprocess");
+        {
+          prm.enter_subsection("Particles");
+          {
+            prm.enter_subsection("Crystal Preferred Orientation");
+            {
+              prm.declare_entry ("Random number seed", "1",
+                                 Patterns::Integer (0),
+                                 "The seed used to generate random numbers. This will make sure that "
+                                 "results are reproducible as long as the problem is run with the "
+                                 "same number of MPI processes. It is implemented as final seed = "
+                                 "user seed + MPI Rank. ");
+
+              prm.declare_entry ("Number of grains per particle", "50",
+                                 Patterns::Integer (1),
+                                 "The number of grains of each different mineral "
+                                 "each particle contains.");
+
+              prm.declare_entry ("Property advection method", "Forward Euler",
```

Comment:
Is ¨Forward Euler" the *best* for this property (whatever that means, but whatever we choose as default will be the method everyone uses). Which method do you usually use?

Reply:
I usually use Backward Euler, so I changed the default to that. I also removed Crank-Nicklson for now (will post about it when I push a new version which also removes the pack and unpack functions)